### PR TITLE
Port to Gtk4

### PR DESCRIPTION
**Pull request by danrabbit**

### CODE REVIEW

File: src/Button.vala [hunk 1]
```diff
@@ -24,19 +24,27 @@ public class PantheonCalculator.Button : Gtk.Button {
...
+        child = lbl;
         tooltip_text = description;
     }
 
-    public override void get_preferred_width (out int minimum_width, out int natural_width) {
-        minimum_width = WIDTH;
-        natural_width = WIDTH;
+    construct {
+        width_request = WIDTH;
+        height_request = HEIGHT;
```
Comment (danrabbit) [inline src/Button.vala:38]: If we want to force aspect ratio for these buttons we should be packing them into a Gtk.AspectFrame instead

File: src/MainWindow.vala [hunk 2]
```diff
@@ -314,9 +301,9 @@ namespace PantheonCalculator {
 
             entry.grab_focus ();
 
-            show_all ();
+            present ();
 
-            this.key_press_event.connect (key_pressed);
+            // key_press_event.connect (key_pressed);
```
Comment (danrabbit) [inline src/MainWindow.vala:306]: Replacement for this is event controllers: https://valadoc.org/gtk4/Gtk.EventControllerKey.html

Comment (Marukesu) [reply src/MainWindow.vala:306]: The Problem that we are using key_press_event to change the keyval in the event, what, i believe, isn't possible with the event controllers.

however, should we be using [Gtk.Editable.insert_text](https://valadoc.org/gtk4/Gtk.Editable.insert_text.html) here?

File: src/Button.vala (hunk 1 above)
Comment (danrabbit) [reply src/Button.vala:34]: Since the window isn't resizable, this doesn't matter

**Review by Marukesu — CHANGES_REQUESTED** [review]
The flatpak manifest is missing an `--device=dri` in the finish-args (required by Gtk4), also, Gtk.Entry now uses delegation, so we need to connect/stop the `insert-text` signal in the `entry.get_delegate()` object.

File: src/HistoryDialog.vala [hunk 3]
```diff
@@ -59,55 +59,66 @@ namespace PantheonCalculator {
...
+            button_clear.add_css_class (Granite.STYLE_CLASS_DESTRUCTIVE_ACTION);
 
             add_button (_("Close"), Gtk.ResponseType.CLOSE);
 
             var button_add = add_button (_("Insert"), Gtk.ResponseType.APPLY);
-            button_add.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
+            button_add.add_css_class (Granite.STYLE_CLASS_SUGGESTED_ACTION);
 
-            show_all ();
+            present ();
```
Comment (Marukesu) [inline src/HistoryDialog.vala:121]: We shouldn't be calling `present()` in the constructor. this make Gtk print this in the terminal:

```
Gtk-Message: 15:20:36.964: GtkDialog mapped without a transient parent. This is discouraged.

(io.elementary.calculator:2): Gtk-WARNING **: 15:20:36.977: gtk_window_set_titlebar() called on a realized window
```

We do set an transient parent, but outside of the constructor. And `set_titlebar()` is called by `Granite.Dialog` in `constructed()`.

File: src/MainWindow.vala [hunk 4]
```diff
@@ -483,11 +468,11 @@ namespace PantheonCalculator {
 
             entry.grab_focus ();
 
-            show_all ();
+            present ();
```
Comment (Marukesu) [inline src/MainWindow.vala:471]: This one don't cause warnings, but, i don't think it's a good thing to call `present()` here too.

**Review by Marukesu — CHANGES_REQUESTED** [review]

**Review by Marukesu — APPROVED** [review]

**Pull request merged**